### PR TITLE
Allow named properties on reactive arrays.

### PR DESCRIPTION
**Pull request by pkaminski**
**What kind of change does this PR introduce?** (check at least one)

- [x] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

If yes, please describe the impact and migration path for existing applications:

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch for v2.x (or to a previous version branch), _not_ the `master` branch
- [ ] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] All tests are passing: https://github.com/vuejs/vue/blob/dev/.github/CONTRIBUTING.md#development-setup
- [x] New/updated tests are included

If adding a **new feature**, the PR's description includes:
- [ ] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

Since arrays can be treated as normal objects in JavaScript and contain a mix of numeric and named properties, this patch makes `Vue.set` and `Vue.del` deal correctly with named properties on reactive arrays.

### CODE REVIEW

File: src/core/observer/index.js [hunk 1]
```diff
@@ -219,11 +219,11 @@ export function set (target: Array<any> | Object, key: any, val: any): any {
  * Delete a property and trigger change if necessary.
  */
 export function del (target: Array<any> | Object, key: any) {
-  if (Array.isArray(target)) {
+  if (Array.isArray(target) && typeof key === 'number') {
     target.splice(key, 1)
     return
   }
-  const ob = target.__ob__
+  const ob = (target : any).__ob__
```
Comment (znck) [inline src/core/observer/index.js]: I can't understand this change. Could you explain?

Comment (pkaminski) [inline src/core/observer/index.js]: I can't say I understand it either, but without this cast Flow complains that "Array doesn't have an
__ob__ property" -- why it decided that `target` is an array, I don't know.  It appears to be an instance of [this Flow issue](https://github.com/facebook/flow/issues/1330), though my knowledge of Flow is minimal so I can't tell for sure.

**Comment by yyx990803** [comment]
I'm not sure if this is a good idea - custom properties on Arrays are ignored during JSON.stringify and cannot be persisted. I personally would never rely on them.

**Comment by pkaminski** [comment]
I'm confused -- what do reactive properties have to do with persistence?  I certainly don't do this often myself, but it's occasionally useful to pass metadata through an API/SPI where you control both ends but can only transmit an array through the library.  IMHO the principle that you can safely replace any member assignment with a call to `Vue.set` outweighs trying to "fix" iffy JavaScript semantics.

**Comment by yyx990803** [comment]
@pkaminski yeah, I guess it doesn't matter that much. I'm just thinking about the implications this has to do with normal persistence scenarios and in particular Vuex snapshots.

**Comment by pkaminski** [comment]
Honestly, I don't even care if the resulting named array property is reactive (though I think it should be for consistency), I'd just like `Vue.set` not to crash when invoked with an array target and a string key.

**Comment by pkaminski** [comment]
Thanks!

**Pull request merged**